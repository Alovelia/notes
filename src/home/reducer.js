import { constants, actions } from 'ducks-helpers';
import handleActions from 'redux-actions/lib/handleActions';
import _ from 'common/helpers';

export const TYPE = constants('home', [
  '~GET',
  '~GET_NOTES',
  //†type
]);
export const ACTION = actions(TYPE);

// The initial state
export const initialState = _.fromJS({
  folders: [],
  notes: [],
});

export default handleActions({
  [TYPE.GET_SUCCESS]: getSuccess,
  [TYPE.GET_NOTES_SUCCESS]: getNotesSuccess,
  //†handler
}, initialState);

function getSuccess(state, { payload }) {
  return state.set('folders', _.fromJS(payload));
}
function getNotesSuccess(state, { payload }) {
  return state.set('notes', _.fromJS(payload));
}
//†reducer
