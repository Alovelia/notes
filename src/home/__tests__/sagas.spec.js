import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';
import { LOCATION_CHANGE } from 'react-router-redux';
import { ACTION, TYPE } from '../reducer';

import {
  getData,
  getDataSaga,
  getNotes,
  getNotesSaga,
} from '../sagas';

describe('getData Saga', () => {
  let getDataGenerator;
  let callDescriptor;
  const folders = [];
  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getDataGenerator = getData();
    callDescriptor = getDataGenerator.next().value;
  });

  it('should call api at first', () => {
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the getSuccess action if it requests the data successfully', () => {
    const response = folders;
    const putDescriptor = getDataGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(ACTION.getSuccess(folders)));
  });

  it('should call the getError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getDataGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(ACTION.getError(response)));
  });
});

describe('getDataSaga Saga', () => {
  const getDataSagaGenerator = getDataSaga();
  const mockedTask = createMockTask();

  it(`should start task to watch for ${TYPE.GET} action`, () => {
    const takeLatestDescriptor = getDataSagaGenerator.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(TYPE.GET, getData));
  });
});

describe('getNotesData Saga', () => {
  let getDataGenerator;
  let callDescriptor;
  const notes = [];
  const action = {
    payload: 'id'
  };
  beforeEach(() => {
    getDataGenerator = getNotes(action);
    callDescriptor = getDataGenerator.next().value;
  });

  it('should call api at first', () => {
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the getNotesSuccess action if it requests the data successfully', () => {
    const response = notes;
    const putDescriptor = getDataGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(ACTION.getNotesSuccess(notes)));
  });

  it('should call the getNotesError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getDataGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(ACTION.getNotesError(response)));
  });
});

describe('getNotesSaga Saga', () => {
  const getDataSagaGenerator = getNotesSaga();
  const mockedTask = createMockTask();

  it(`should start task to watch for ${TYPE.GET_NOTES} action`, () => {
    const takeLatestDescriptor = getDataSagaGenerator.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(TYPE.GET_NOTES, getNotes));
  });
});
