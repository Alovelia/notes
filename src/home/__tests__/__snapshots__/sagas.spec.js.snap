// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getData Saga should call api at first 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/folders",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getNotesData Saga should call api at first 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "/folders/id/notes",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
